src/valkeymodule.h

```diff
@@ -794,6 +795,91 @@ typedef void (*ValkeyModuleInfoFunc)(ValkeyModuleInfoCtx *ctx, int for_crash_rep
 typedef void (*ValkeyModuleDefragFunc)(ValkeyModuleDefragCtx *ctx);
 typedef void (*ValkeyModuleUserChangedFunc)(uint64_t client_id, void *privdata);
 
+/* Current ABI version for scripting engine modules. */
+#define VALKEYMODULE_SCRIPTING_ENGINE_ABI_VERSION 1UL
+
+/* Type definitions for implementing scripting engines modules. */
+typedef void ValkeyModuleScriptingEngineCtx;
+typedef void ValkeyModuleScriptingEngineFunctionCtx;
+
+/* This struct represents a scripting engine function that results from the
+ * compilation of a script by the engine implementation.
+ *
+ * IMPORTANT: If we ever need to add/remove fields from this struct, we need
+ * to bump the version number defined in the
+ * `VALKEYMODULE_SCRIPTING_ENGINE_ABI_VERSION` constant.
+ */
+typedef struct ValkeyModuleScriptingEngineCompiledFunction {
+    ValkeyModuleString *name; /* Function name */
+    void *function;           /* Opaque object representing a function, usually it'
+                                 the function compiled code. */
+    ValkeyModuleString *desc; /* Function description */
+    uint64_t f_flags;         /* Function flags */
+} ValkeyModuleScriptingEngineCompiledFunction;
+
+typedef ValkeyModuleScriptingEngineCompiledFunction **(*ValkeyModuleScriptingEngineCreateFunctionsLibraryFunc)(
+    ValkeyModuleCtx *module_ctx,
+    ValkeyModuleScriptingEngineCtx *engine_ctx,
+    const char *code,
+    size_t timeout,
+    size_t *out_num_compiled_functions,
+    char **err);
```

Comment:
Thinking about this some more, I'd like to propose returning an integer error code instead:


```suggestion
typedef int (*ValkeyModuleScriptingEngineCreateFunctionsLibraryFunc)(
    ValkeyModuleCtx *module_ctx,
    ValkeyModuleScriptingEngineCtx *engine_ctx,
    const char *code,
    size_t timeout,
    size_t *out_num_compiled_functions,
    ValkeyModuleScriptingEngineCompiledFunction **compiled_functions);
```

This would

1. avoid the assumption that a scripting module would allocate these strings using zmalloc, which can't be enforced by compiler
2. eliminate the "char * to sds" conversion
3. help ensure consistency over the error strings across different scripting engines.

Reply:
I  like the suggestion but I see one problem with 3. How the scripting engine communicates the kind of error that made this function to fail back to the client? 

For instance, most of the times this function might fail because of a syntax error in the script, and the scripting engine can give a nice message specifying where the syntax error is located, which in this case will not be sent to the client.

We might log the error in the server log, but I assume the client of valkey might not have access to the server logs.

Reply:
Interesting. Do you have an example? How does the built-in Lua interpreter handle syntax errors? I wonder if we could generalize an outgoing line number and an outgoing column number? Or we'd like to provide more detailed context like a real compiler.

Reply:
Currently, when there's some syntax, or runtime error we send the error message returned by lua:
```
127.0.0.1:6379> function load replace  "#!lua name=mylib\nlocal function foo()\nretrn 1\nend\nserver.register_function('foo', bar)\n"
(error) ERR Error compiling function: user_function:3: '=' expected near '1'

127.0.0.1:6379> function load replace  "#!lua name=mylib\nlocal function foo()\nreturn 1\nend\nserver.register_function('foo', bar)\n"
(error) ERR Error registering functions: ERR user_function:5: Script attempted to access nonexistent global variable 'bar'
```

I think the more generic approach is to allow the scripting engine to return a C string with the error message.

Reply:
@PingXie How about we get this PR merged and we can still change the API a bit more before 8.1 is released? Ricardo has the follow-up PR for EVAL waiting to be added. It will slightly modify the API already.

Reply:
@rjd15372 We should document this in some way for the module API docs, for example that the module needs to allocate the `err` using `ValkeyModule_Alloc` (a wrapper for zmalloc).

The comments before each `VM_` function definition gets extracted to produce this page: https://valkey.io/topics/modules-api-ref/

We also put comments for structs in valkeymodule.h, but it's not perfect for documentatios. For example, Rust module authors don't directly use this file.

We can consider creating a new documentation topic page for engines later.

Reply:
> How about we get this PR merged and we can still change the API a bit more before 8.1 is released?

Agreed.

Reply:
> I think the more generic approach is to allow the scripting engine to return a C string with the error message.

Yeah if we would like to provide detailed compilation errors then we need to keep this capability but I would suggest `robj` too over C strings.